From 067cf4546a37cf7f13d6a34e252ef7b8ee1264b0 Mon Sep 17 00:00:00 2001
From: lirui <lirui@unknown>
Date: Fri, 6 Mar 2015 05:55:45 +0000
Subject: [PATCH 0070/1164] HIVE-9860: MapredLocalTask/SecureCmdDoAs leaks
 local files (Brock via Rui)

git-svn-id: https://svn.apache.org/repos/asf/hive/trunk@1664548 13f79535-47bb-0310-9956-ffa450edef68
(cherry picked from commit c2a847903d85a0d982a63bdc7788134687dca3c8)
---
 .../apache/hadoop/hive/ql/exec/SecureCmdDoAs.java  |    8 ++++++--
 .../hadoop/hive/ql/exec/mr/MapredLocalTask.java    |   10 +++++++---
 2 files changed, 13 insertions(+), 5 deletions(-)

diff --git a/ql/src/java/org/apache/hadoop/hive/ql/exec/SecureCmdDoAs.java b/ql/src/java/org/apache/hadoop/hive/ql/exec/SecureCmdDoAs.java
index 619aa1f..15716fc 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/exec/SecureCmdDoAs.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/exec/SecureCmdDoAs.java
@@ -38,6 +38,7 @@
  */
 public class SecureCmdDoAs {
   private final Path tokenPath;
+  private final File tokenFile;
 
   public SecureCmdDoAs(HiveConf conf) throws HiveException, IOException{
     // Get delegation token for user from filesystem and write the token along with
@@ -46,8 +47,8 @@ public SecureCmdDoAs(HiveConf conf) throws HiveException, IOException{
     FileSystem fs = FileSystem.get(conf);
     Token<?> fsToken = fs.getDelegationToken(uname);
 
-    File t = File.createTempFile("hive_hadoop_delegation_token", null);
-    tokenPath = new Path(t.toURI());
+    tokenFile = File.createTempFile("hive_hadoop_delegation_token", null);
+    tokenPath = new Path(tokenFile.toURI());
 
     //write credential with token to file
     Credentials cred = new Credentials();
@@ -60,4 +61,7 @@ public void addEnv(Map<String, String> env){
         tokenPath.toUri().getPath());
   }
 
+  public void close() {
+    tokenFile.delete();
+  }
 }
diff --git a/ql/src/java/org/apache/hadoop/hive/ql/exec/mr/MapredLocalTask.java b/ql/src/java/org/apache/hadoop/hive/ql/exec/mr/MapredLocalTask.java
index 9f3df99..d36d003 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/exec/mr/MapredLocalTask.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/exec/mr/MapredLocalTask.java
@@ -97,6 +97,7 @@
   private ExecMapperContext execContext = null;
 
   private Process executor;
+  private SecureCmdDoAs secureDoAs;
 
   public MapredLocalTask() {
     super();
@@ -271,7 +272,7 @@ public int executeInChildVM(DriverContext driverContext) {
         //If kerberos security is enabled, and HS2 doAs is enabled,
         // then additional params need to be set so that the command is run as
         // intended user
-        SecureCmdDoAs secureDoAs = new SecureCmdDoAs(conf);
+        secureDoAs = new SecureCmdDoAs(conf);
         secureDoAs.addEnv(variables);
       }
 
@@ -314,9 +315,12 @@ public int executeInChildVM(DriverContext driverContext) {
 
       return exitVal;
     } catch (Exception e) {
-      e.printStackTrace();
-      LOG.error("Exception: " + e.getMessage());
+      LOG.error("Exception: " + e, e);
       return (1);
+    } finally {
+      if (secureDoAs != null) {
+        secureDoAs.close();
+      }
     }
   }
 
-- 
1.7.9.5

